# Numerical Integration

In this tutorial we describe how to use Equadratures capabilities to numerically compute integrals using efficient quadratures.

In [13]:
import numpy as np
import matplotlib.pyplot as plt
import equadratures as eq


order = 10
parameter = eq.Parameter(lower=-1, upper=1, order=order, distribution='uniform')

We will first consider univariate polynomials, evaluating using Gauss-Legendre Quadrature.

Above, we being by defining a uniform paramter from -1 to 1 and of order 10, meaning that we are only considering the first 10 orthogonal polynomials in our quadrature computations.

$$
\int_{-1}^{1}f(x)dx
$$

Utilizing Gauss-Legendre Quadrature (since we are using a uniform weight distribution), we can evaluate the above integral by evaluating the function at certain points, that correspond to the roots of the Legendre polynomial, and dot those evaluations with weights associated with those points.

In [14]:
basis = eq.Basis('univariate')
poly = eq.Poly(parameter, basis, method='numerical-integration')
points, weights = poly.get_points_and_weights()

Equadratures allows simple retrieval of these weights and points, allowing us to evaluate computationally efficient integrals.

$$
\int_{-1}^{1}f(x)dx \approx \sum_{n=1}^{i}f(points_i)*weights_i 
$$

In [15]:
def f(x):
    return np.cos(x**2);

Consider the function:
$$
f(x) = cos(x^2)
$$


Utilizing the weights and points we previously collected, evaluation is easily done through the following.

In [16]:
eval = float( 2  * np.dot(weights , eq.evaluate_model(points, f) ) )
print(eval)

1.809048475800536


$$
\int_{-1}^{1}cos(x^2)dx \approx \sum_{n=1}^{i}f(points_i)*weights_i = 1.809048475800536
$$

This approach can be altered to help compute a wide variety of integrals. Including switching away from Gauss-Legendre, by specifying a non uniform distribution.


# Bivariate Distributions

Given that in the real world, we are typically interested in functions of multiple variables.

Equadratures allows this through the specification of the multi-index, the possible multi-indices are called index sets.

The following are the 6 index sets in Equadratures: tensor-grid, total-order, sparse-grid, hyperbolic-basis, and euclidean degree.

For the pruposes of this tutorial, we shall use a tensor grid.

In [17]:
tensor = eq.Basis('tensor-grid', [1,1])
biParam = eq.Parameter(lower=-1, upper=1, order=order, distribution='Uniform')

biPoly = eq.Poly([biParam, biParam], tensor, method='numerical-integration')

biPoints, biWeights = biPoly.get_points_and_weights()

Now consider the multivariable function:
$$
g(x) = \int_{-1}^{1}\int_{-1}^{1}16x^3 + 8y^2 - 8y+6 \;dxdydz
$$

In [18]:
def g(x):
    return 16*x[0]**3+8*x[1]**2-8*x[1]+6;

xValuesBi = np.linspace(-1., 1., 1000)
yValuesBi = np.linspace(-1., 1., 1000)
zValuesBi = np.linspace(-1., 1., 1000)

print(biPoints)
print(biPoints[:,0])
print(biPoints[:,1])

[[-0.57735027 -0.57735027]
 [-0.57735027  0.57735027]
 [ 0.57735027 -0.57735027]
 [ 0.57735027  0.57735027]]
[-0.57735027 -0.57735027  0.57735027  0.57735027]
[-0.57735027  0.57735027 -0.57735027  0.57735027]


The numerical integral is now easily calculated using the points nad weights we had earlier computed.

In [19]:
evalBi = float( 2  * np.dot(biWeights , eq.evaluate_model(biPoints, g) ) )
print(evalBi);

17.333333333333325


$$
g(x) = \int_{-1}^{1}\int_{-1}^{1}16x^3 + 8y^2 - 8y+6 \;dxdydz = 17.\overline{33}
$$